### PR TITLE
coprocessor: add RPN functions 'cast * as int'

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

## What have you changed? (mandatory)

Add CAST * as int RPN functions

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- New feature (change which adds functionality)

### CODE REVIEW

File: src/coprocessor/dag/rpn_expr/impl_cast.rs [hunk 1]
```diff
@@ -115,6 +173,125 @@ pub fn cast_int_as_decimal(
...
+            $as_uint_fn,
+            $as_int_conv,
+            $as_uint_conv,
+            $extra
+        );
+    };
+    (_inner, $ty:ty, $as_int_fn:ident, $as_uint_fn:ident, $as_int_conv:ident, $as_uint_conv:ident, $expr:expr) => {
+        #[rpn_fn(capture = [ctx, extra])]
+        #[inline]
+        pub fn $as_int_fn(
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_cast.rs]: how about the duplicate code when other types are going to be implemented? (e.g. cast_xxx_as_real)

Comment (lonng) [reply src/coprocessor/dag/rpn_expr/impl_cast.rs]: I think the ultimate solution is the procedural macro (at least current macro cannot apply to cast_xxx_as_real). 🤔

File: src/coprocessor/dag/rpn_expr/impl_cast.rs [hunk 2]
```diff
@@ -33,6 +34,63 @@ pub fn get_cast_fn_rpn_node(
         (EvalType::DateTime, EvalType::Real) => cast_time_as_real_fn_meta(),
         (EvalType::Duration, EvalType::Real) => cast_duration_as_real_fn_meta(),
         (EvalType::Json, EvalType::Real) => cast_json_as_real_fn_meta(),
+        (EvalType::Int, EvalType::Int) => {
```
Comment (iosmanthus) [inline src/coprocessor/dag/rpn_expr/impl_cast.rs]: Could we use `match` instead?

Comment (lonng) [reply src/coprocessor/dag/rpn_expr/impl_cast.rs]: The `i` version will generate fewer ASM instructions than the `m1` version.
```asm
playground::m1:
	subq	$32, %rsp
	movb	%sil, %al
	movb	%dil, %cl
	andb	$1, %cl
	movb	%cl, 6(%rsp)
	andb	$1, %al
	movb	%al, 7(%rsp)
	movb	6(%rsp), %al
	movb	7(%rsp), %cl
	andb	$1, %al
	movb	%al, 24(%rsp)
	andb	$1, %cl
	movb	%cl, 25(%rsp)
	testb	$1, 24(%rsp)
	jne	.LBB59_5
	jmp	.LBB59_6
```
```asm
playground::i:
	subq	$24, %rsp
	movb	%sil, %al
	movb	%dil, %cl
	andb	$1, %cl
	movb	%cl, 6(%rsp)
	andb	$1, %al
	movb	%al, 7(%rsp)
	testb	$1, 6(%rsp)
	je	.LBB61_6
	testb	$1, 7(%rsp)
	je	.LBB61_5
	leaq	.L__unnamed_6(%rip), %rax
	movq	%rax, 8(%rsp)
	movq	$10, 16(%rsp)
```

File: src/coprocessor/dag/rpn_expr/impl_cast.rs [hunk 3]
```diff
@@ -33,6 +34,63 @@ pub fn get_cast_fn_rpn_node(
...
+            if !from_field_type.is_unsigned() {
+                if !to_field_type.is_unsigned() {
+                    cast_int_as_int_fn_meta()
```
[the remark this answers is not in the document]
Comment (lonng) [reply src/coprocessor/dag/rpn_expr/impl_cast.rs]: Just keep the same style with the previous code.

File: src/coprocessor/dag/rpn_expr/impl_cast.rs (hunk 2 above)
Comment (lonng) [reply src/coprocessor/dag/rpn_expr/impl_cast.rs]: Anyway, I have changed to the `match` patter for code readability.

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/impl_cast.rs]: Both styles look good to me..

**Pull request merged**